Lib/test/test_builtin.py

```diff
@@ -491,6 +485,29 @@ async def arange(n):
         finally:
             asyncio.set_event_loop_policy(policy)
 
+    def test_top_level_async_in_assert_compiles(self):
+        # gh-121637: do the right thing when async code is optimized away
+        modes = ('single', 'exec')
+        optimizations = (0, 1, 2)
+        code_samples = [
+            '''assert not await asyncio.sleep(0)''',
+            '''assert [x async for x in arange(1)]''',
+            '''assert {x async for x in arange(1)}''',
+            '''assert {x: x async for x in arange(1)}''',
+        ]
+        for mode, code_sample, optimize in product(modes, code_samples, optimizations):
+            with self.subTest(mode=mode, code_sample=code_sample, optimize=optimize):
+                source = dedent(code_sample)
+                if (optimize):
+                    co = compile(source, '?', mode, optimize=optimize)
+                    co_expected = compile('pass',  '?', mode, optimize=optimize)
+                    self.assertEqual(list(co.co_lines()), list(co_expected.co_lines()))
```

Comment:
This looks like we no longer raise an error in optimized mode if there is an invalid await. That seems wrong and doesn't align with what we decided on in #121637.

Reply:
If we want to stick with this then we probably can't optimise the assertions in ast_opt.